### PR TITLE
HPCC-27724 Minor improvements to udp buffer handling

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: roxie/roxiemem/roxiemem.cpp [hunk 1]
```diff
@@ -4823,34 +4842,61 @@ class CChunkingRowManager : public CRowManager
...
-                    finger->nextActive = 0;
-                    finger->Release();
-                    dataBuffs--;
-                    possibleGoers.fetch_sub(1, std::memory_order_relaxed); // It doesn't matter when other threads see this update
-                    if (last)
-                        last->nextActive = next;
+                    assertex(finger->mgr == this);
+                    // MORE - if we get a load of data in and none out this can start to bog down...
+                    DataBuffer *next = finger->nextActive;
+                    assertex(getSeqFromDataId((memsize_t)next) == 0);
```
Comment (richardkchapman) [inline roxie/roxiemem/roxiemem.cpp:4863]: Earlier you logged this anomaly then continued - here you throw an exception...

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Some questions and minor niggles

File: roxie/roxiemem/roxiemem.cpp [hunk 2]
```diff
@@ -4459,18 +4462,34 @@ class CChunkingRowManager : public CRowManager
...
 
         DataBuffer *dfinger = activeBuffs;
+        unsigned gone =0;
+        unsigned dead = 0;
         while (dfinger)
         {
             DataBuffer *next = dfinger->nextActive;
-            if (memTraceLevel >= 2 && dfinger->queryCount()!=1)
+            if (getSeqFromDataId((memsize_t)next) != 0)
+                logctx.CTXLOG("INTERNAL ERROR: unexpected sequence number in nextActive in ~CChunkingRowManager - addr=%p rowMgr=%p",
```
[the remark this answers is not in the document]
Comment (mckellyln) [reply roxie/roxiemem/roxiemem.cpp:4472]: This is resolved with the memTraceInconsistencies var, correct ?

File: roxie/roxiemem/roxiemem.cpp [hunk 3]
```diff
@@ -4459,18 +4463,34 @@ class CChunkingRowManager : public CRowManager
...
-            dfinger->nextActive = 0;
-            dfinger->mgr = NULL; // Avoid calling back to noteDataBufferReleased, which would be unhelpful
+            dfinger->nextActive = nullptr;
+            dfinger->mgr = nullptr; // Avoid calling back to noteDataBufferReleased, which would be unhelpful
             dfinger->count.store(0, std::memory_order_relaxed);
             dfinger->released();
             dfinger = next;
         }
+
+        if ((dataBuffs != possibleGoers.load()) || (gone != possibleGoers))
```
Comment (mckellyln) [inline roxie/roxiemem/roxiemem.cpp:4491]: Perhaps also add && if (memTraceInconsistencies) to the logging here ?

File: roxie/roxiemem/roxiemem.cpp [hunk 4]
```diff
@@ -4823,34 +4842,61 @@ class CChunkingRowManager : public CRowManager
...
+                    possibleGoers.fetch_sub(gone, std::memory_order_relaxed); // It doesn't matter when other threads see this update
+                    dataBuffs -= gone;
                 }
-                else
-                    last = finger;
-                finger = next;
             }
+
             assert(dataBuff->nextActive == nullptr);
+            dataBuff->filler = ++attachSeq;
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply roxie/roxiemem/roxiemem.cpp:4904]: no.  It might be useful for tracking packets in a debugger - changed to debug only

**Review by richardkchapman — APPROVED** [review]

**Review by mckellyln — COMMENTED** [review]
Just one comment / question

File: roxie/roxiemem/roxiemem.cpp (hunk 3 above)
Comment (ghalliday) [reply roxie/roxiemem/roxiemem.cpp:4491]: I think this should only happen once per query, and be very unlikely - so I think it is ok to trace it always.  The other tracing could be generated many times per query.

Comment (mckellyln) [reply roxie/roxiemem/roxiemem.cpp:4491]: ok.
Approved.

**Review by mckellyln — APPROVED** [review]
Looks good to merge.

**Pull request merged**